Fix Search Results Count Not Being Returned

interlineae opened the pull request:
## Proposed changes:

Currently, all search results return the string `Showing popular results`, but we would instead expect to see `Found X results for $search_term`. This PR proposes a change to logic that ensures this expected string is not dependent upon a check for `corrected_query`.

Namely, the issue is that `if ( hasQuery && hasCorrectedQuery ) {` was returning `false` when searching for correctly spelt words. This bugged behaviour can be seen when running searches on https://jetpack.com.

(I'm not sure if I did the patch files right or wrong, or if I should have answered `false` to the "would a user / theme author like to know this?" question, so do let me know.)

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

N/A

## Does this pull request change what data or activity we track or use?

No

## Testing instructions:

In your local environment/environment exposed through `ngrok` or `Jurassic.tube`:

- While still on `trunk`, run an instant search to confirm the `Showing popular results` text
- Switch to this branch (`fix/search-results-count`) and run another instant search to confirm that `Found X result for $search_term` displays the correct information.
- Can also test on WPcom / WoA site to see the same result.

Expected results:

- On initial load, `Showing popular results`.
- When searching with 1 returned result, `Found 1 result for...`.
- When searching with more than 1 returned result, `Found X results for...`.
- When searching nonsense with no correction, `No results found`.
- When searching a misspelling with a correction, `Found 1 result for...` or `Found X results for...`.

interlineae committed:
Fix search result count

This commit fixes a bug with how search results are displayed. Currently,
combinging hasQuery with hasCorrectedQuery causes the if that was on line
70 to return false, which triggers the default return of the function, around
line 118 with "Showing popular results" (which should just be a default state
on first load, etc.).

This adds "smarter" logic handling so that we show the correct "Founx X result/s"

interlineae committed:
changelog

interlineae committed:
Update fix-search-results-count

trakos reviewed (commented):
Nice find! From what I can tell, the removal of search results count was an unintended consequence of PR https://github.com/Automattic/jetpack/pull/17663.

Maybe it should be a separate `else if ( has Query ) {` after the `isMultiSite` check? You can see a different message for multisites when using "search in all Automattic" on A8C p2-- this would stop working for queries.

Minor note: we don't really need the "for "%2$s" when query isn't corrected, so maybe we could just use `Found %s results` for queries without correction. That way, the corrections would stand out more.

What do you think?

> (I'm not sure if I did the patch files right or wrong, or if I should have answered false to the "would a user / theme author like to know this?" question, so do let me know.)

Mentioning it in the notes makes sense to me. I'd maybe prefix the message with `Search: `.

interlineae committed:
Move hasQuery check below isMultiSite check

- Move the hasQuery check below the isMultiSite check so that we run the latter
first, otherwise, it will never be reached if hasQuery is first.

- Minor update to `_n()` output when there is no corrected query.

interlineae committed:
Merge branch 'fix/search-results-count' of a8c.github.com:Automattic/jetpack into fix/search-results-count

interlineae committed:
Update fix-search-results-count

interlineae committed:
Update changelog message

anaemnesis commented:
> Nice find! From what I can tell, the removal of search results count was an unintended consequence of PR #17663.
> 
> Maybe it should be a separate `else if ( has Query ) {` after the `isMultiSite` check? You can see a different message for multisites when using "search in all Automattic" on A8C p2: https://jetpackdatap2.wordpress.com/?s=test&group_id=p2_workspace_hub_blog_id%3A188873359 - this would stop working for queries.
> 
> Minor note: we don't really need the "for "%2$s" when query isn't corrected, so maybe we could just use `Found %s results` for queries without correction. That way, the corrections would stand out more.
> 
> What do you think?
> 
> > (I'm not sure if I did the patch files right or wrong, or if I should have answered false to the "would a user / theme author like to know this?" question, so do let me know.)
> 
> Mentioning it in the notes makes sense to me. I'd maybe prefix the message with `Search: `.

Good shout.

I've moved the `hasQuery` check to after `isMultiSite`. I also looked into options for for the 4 `sprintf` returns as they're all fairly similar, but I've decided to leave these as-is rather than throw them into a new function to return, etc.

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into fix/search-results-count

trakos reviewed (approved):
Thank you for the updates, looks good to me!

interlineae committed:
Merge remote-tracking branch 'origin/trunk' into fix/search-results-count
